tooling/hcpctl/pkg/breakglass/minting/minting_test.go

```diff
@@ -0,0 +1,505 @@
+// Copyright 2025 Microsoft Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package minting
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"testing"
+	"time"
+
+	certificatesv1alpha1 "github.com/openshift/hypershift/api/certificates/v1alpha1"
+	certificatesv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	kubefake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/rest"
+	ktesting "k8s.io/client-go/testing"
+	client "sigs.k8s.io/controller-runtime/pkg/client"
+	clientfake "sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	"github.com/Azure/ARO-HCP/tooling/hcpctl/pkg/common"
+)
+
+func TestNewDefaultManager(t *testing.T) {
+	restConfig := &rest.Config{Host: "https://test-server"}
+
+	manager, err := NewDefaultManager(restConfig)
+	if err != nil {
+		t.Fatalf("Expected no error, got %v", err)
+	}
+
+	if manager == nil {
+		t.Error("NewDefaultManager returned nil")
+		return
+	}
+
+	if manager.kubeClient == nil {
+		t.Error("kubeClient not set")
+	}
+
+	if manager.ctrlClient == nil {
+		t.Error("ctrlClient not set")
+	}
+}
+
+func TestCreateCSR(t *testing.T) {
+	tests := []struct {
+		name              string
+		csrName           string
+		csrPEM            []byte
+		signerName        string
+		labels            map[string]string
+		expirationSeconds int32
+		expectError       bool
+		setupClient       func(*kubefake.Clientset)
+	}{
+		{
+			name:              "successful CSR creation",
+			csrName:           "test-csr",
+			csrPEM:            []byte("fake-csr-pem"),
+			signerName:        "test-signer",
+			labels:            map[string]string{"test": "label"},
+			expirationSeconds: 3600,
+			expectError:       false,
+			setupClient:       func(client *kubefake.Clientset) {},
+		},
+		{
+			name:              "CSR creation failure",
+			csrName:           "test-csr",
+			csrPEM:            []byte("fake-csr-pem"),
+			signerName:        "test-signer",
+			labels:            map[string]string{"test": "label"},
+			expirationSeconds: 3600,
+			expectError:       true,
+			setupClient: func(client *kubefake.Clientset) {
+				client.PrependReactor("create", "certificatesigningrequests", func(action ktesting.Action) (handled bool, ret runtime.Object, err error) {
```

Comment:
Strongly (x10000000000000000) recommend not using fake clients, any of this stuff, etc. It's super brittle. Recommend writing an end-to-end test (running the controller isn't hard) or factoring your code as a functional core surrounded by imperative shell, so you can assert over "given this state, what do you do next"?